db/active_replicator_checkpointer.go

```diff
@@ -177,42 +159,180 @@ func calculateCheckpointSeq(expectedSeqs []string, processedSeqs map[string]stru
 		return
 	}
 
-	seq = expectedSeqs[maxI]
+	seq = c.expectedSeqs[maxI]
 
-	if len(expectedSeqs)-1 == maxI {
+	if len(c.expectedSeqs)-1 == maxI {
 		// received full set, empty list
-		expectedSeqs = expectedSeqs[0:0]
+		c.expectedSeqs = c.expectedSeqs[0:0]
 	} else {
 		// trim sequence list for partially received set
-		expectedSeqs = expectedSeqs[maxI+1:]
+		c.expectedSeqs = c.expectedSeqs[maxI+1:]
 	}
 
 	return seq
 }
 
-func (c *Checkpointer) setCheckpoint(seq string) {
+const checkpointDocIDPrefix = "checkpoint/"
+
+// fetchCheckpoints tries to fetch since values for the given replication by requesting a checkpoint on the local and remote.
+// If we fetch missing, or mismatched checkpoints, we'll pick the lower of the two, and roll back the higher checkpoint.
+func (c *Checkpointer) fetchCheckpoints() error {
+	base.TracefCtx(c.ctx, base.KeyReplicate, "fetchCheckpoints()")
+
+	localSeq, localRev, err := c.getLocalCheckpoint()
+	if err != nil {
+		return err
+	}
+	base.DebugfCtx(c.ctx, base.KeyReplicate, "got local checkpoint: %q %q", localSeq, localRev)
+	c.lastLocalCheckpointRevID = localRev
+
+	remoteSeq, remoteRev, err := c.getRemoteCheckpoint()
+	if err != nil {
+		return err
+	}
+	base.DebugfCtx(c.ctx, base.KeyReplicate, "got remote checkpoint: %q %q", remoteSeq, remoteRev)
+	c.lastRemoteCheckpointRevID = remoteRev
+
+	// If localSeq and remoteSeq match, we'll use this value.
+	checkpointSeq := localSeq
+
+	// Determine the lowest sequence if they don't match
+	if localSeq != remoteSeq {
+		base.DebugfCtx(c.ctx, base.KeyReplicate, "sequences mismatched, finding lowest of %q %q", localSeq, remoteSeq)
+		localSeqVal, err := parseIntegerSequenceID(localSeq)
+		if err != nil {
+			return err
+		}
+
+		remoteSeqVal, err := parseIntegerSequenceID(remoteSeq)
+		if err != nil {
+			return err
+		}
+
+		// roll local/remote checkpoint back to lowest of the two
+		if remoteSeqVal.Before(localSeqVal) {
+			checkpointSeq = remoteSeq
+			c.lastLocalCheckpointRevID, err = c.setLocalCheckpoint(checkpointSeq, c.lastLocalCheckpointRevID)
+			if err != nil {
+				return err
+			}
+		} else {
+			checkpointSeq = localSeq
+			c.lastRemoteCheckpointRevID, err = c.setRemoteCheckpoint(checkpointSeq, c.lastRemoteCheckpointRevID)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	if checkpointSeq == "" {
+		c.StatGetCheckpointMissTotal.Add(1)
+	} else {
+		c.StatGetCheckpointHitTotal.Add(1)
+	}
+
+	base.DebugfCtx(c.ctx, base.KeyReplicate, "using checkpointed seq: %q", checkpointSeq)
+	c.lastCheckpointSeq = checkpointSeq
+
+	return nil
+}
+
+func (c *Checkpointer) setCheckpoints(seq string) error {
```

Comment:
I'm wondering about the conflict error handling here.  In the event of an error during normal checkpointing, we just log a warning and carry on.  In the case of a conflict (maybe a case where two nodes are running the same replication because of some heartbeat-related race), will the replication just always fail to write checkpoints from that point forward, due to rev mismatch?  It feels like we should reload/reset the checkpoint when we fail to write it due to conflict.